### PR TITLE
Forces to thruster signal

### PR DESCRIPTION
**Pull request by joaobrittoneto**
Convert thrusters forces to thruster's command signal, according to its dynamics. 
To be used between AccelerationController and thruster's driver.

### CODE REVIEW

File: tasks/ThrustersInput.cpp [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    		// Well, if the thruster_coeff vectors have the same size, looks like the control_modes is the one with wrong size
+    		exception(WRONG_SIZE_OF_CONTROLMODES);
+    	}
+    	else if(controlModes.empty())
+    	{
+    		// In this case, control_modes is empty and coeffPos has the right number of thrusters
+    		// Let's set control_modes with default values RAW
+    		numberOfThrusters = coeffPos.size();
+        	//resize the controlModes to number of Thrusters
+        	controlModes.resize(numberOfThrusters);
```
Comment (doudou) [inline tasks/ThrustersInput.cpp]: resize() can be provided with the initialization value

```
controlModes.resize(numberOfThrusters, base::JointState::RAW)
```

File: tasks/ThrustersInput.cpp [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        	}
+    	}
+    }
+
+    // Check control_modes
+    for (int i=0; i<numberOfThrusters; i++)
+    {
+    	if(controlModes[i] !=  base::JointState::SPEED && controlModes[i] !=  base::JointState::RAW)
+    	{
+			RTT::log(RTT::Error) << "Control Mode should be SPEED or RAW" << RTT::endlog();
```
Comment (doudou) [inline tasks/ThrustersInput.cpp]: Wrong indentation, you're mixing tabs and spaces and have your editor configured to do tab == 4 spaces.

Use tab==8 and indent with spaces.

File: tasks/ThrustersInput.cpp [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    {
+		RTT::log(RTT::Error) << "Thruster coefficients were not set" << RTT::endlog();
+    	exception(WRONG_SIZE_OF_THRUSTER_COEFFICIENTS);
+    }
+
+
+    // Check size of parameters. They should met among them.
+    if(coeffPos.size() != numberOfThrusters || coeffNeg.size() != numberOfThrusters )
+    {
+    	if(coeffPos.size() != coeffNeg.size())
```
Comment (doudou) [inline tasks/ThrustersInput.cpp]: This makes the whole logic hard to read, and I believe will be hard to understand on the usage side as well. All three have to agree, so test for that and output a generic exception (THRUSTER_ARRAYS_SIZE_DIFFER), message

```
The size of control_modes, thruster_coeff_pos and thruster_coeff_neg do not agree (... output the size of the 3 ...)
```

File: tasks/ThrustersInput.cpp [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    	}
+    }
+
+    // Check control_modes
+    for (int i=0; i<numberOfThrusters; i++)
+    {
+    	if(controlModes[i] !=  base::JointState::SPEED && controlModes[i] !=  base::JointState::RAW)
+    	{
+			RTT::log(RTT::Error) << "Control Mode should be SPEED or RAW" << RTT::endlog();
+		exception(WRONG_SET_OF_CONTROLMODES);
```
Comment (doudou) [inline tasks/ThrustersInput.cpp]: return false if there is a configuration problem.

File: tasks/ThrustersInput.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    {
+    	if(coeffPos.size() != coeffNeg.size())
+    	{	// If coefficients sizes are different among them and controm_modes size, one of the coeff were not set right
+    		RTT::log(RTT::Error) << "Coefficients were not set with "<< numberOfThrusters <<" as predicted" << RTT::endlog();
+    		exception(WRONG_SIZE_OF_THRUSTER_COEFFICIENTS);
+    	}
+    	else if(!controlModes.empty())
+    	{
+    		// Well, if the thruster_coeff vectors have the same size, looks like the control_modes is the one with wrong size
+    		exception(WRONG_SIZE_OF_CONTROLMODES);
```
Comment (doudou) [inline tasks/ThrustersInput.cpp]: return false if there is a configuration problem

File: tasks/ThrustersInput.cpp [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+
+    // Check size of parameters. They should met among them.
+    if(coeffPos.size() != numberOfThrusters || coeffNeg.size() != numberOfThrusters )
+    {
+    	if(coeffPos.size() != coeffNeg.size())
+    	{	// If coefficients sizes are different among them and controm_modes size, one of the coeff were not set right
+    		RTT::log(RTT::Error) << "Coefficients were not set with "<< numberOfThrusters <<" as predicted" << RTT::endlog();
+    		exception(WRONG_SIZE_OF_THRUSTER_COEFFICIENTS);
```
Comment (doudou) [inline tasks/ThrustersInput.cpp]: return false if there is a configuration problem

File: tasks/ThrustersInput.cpp [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    controlModes		= _control_modes.get();
+
+    // Let's assume control_modes was set with the right number of thrusters
+    numberOfThrusters = controlModes.size();
+
+    // Those coefficients need to be set
+    if(coeffPos.empty() || coeffNeg.empty())
+    {
+		RTT::log(RTT::Error) << "Thruster coefficients were not set" << RTT::endlog();
+    	exception(WRONG_SIZE_OF_THRUSTER_COEFFICIENTS);
```
Comment (doudou) [inline tasks/ThrustersInput.cpp]: return false if there is a configuration problem

File: tasks/ThrustersInput.cpp [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+		RTT::log(RTT::Error) << "The input has not "<< numberOfThrusters <<
+				" thruster as predicted. Actual size is "<< cmd_in.elements.size() << ". Check configuration. "<< RTT::endlog();
+		exception(UNEXPECTED_THRUSTER_INPUT);
+		return false;
+	}
+
+	for (int i=0; i<numberOfThrusters; i++)
+	{
+		// Define how to call the problematic thruster
+		std::string textThruster;
```
Comment (doudou) [inline tasks/ThrustersInput.cpp]: Don't do this computation if you don't need it, i.e. move it to the if (ERROR) { } below

File: tasks/ThrustersInput.cpp [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+		if(cmd_in.size() == numberOfThrusters)
```
Comment (doudou) [inline tasks/ThrustersInput.cpp]: Isn't this always going to be true ? You check for it above ...

**Comment by rafaelsaback** [comment]
Corrected the error checking and improved the code as a whole. I also used some unitary tests to check the configuration exceptions and also the calculated values.

**Comment by rafaelsaback** [comment]
I've just realized that we still need to implement limits for the output. To be done soon...

**Comment by rafaelsaback** [comment]
@doudou, added the missing commit here

**Comment by saarnold** [comment]
One could also allow efforts and just hand them through, but it's not necessary.
Is this task complete, can it be merged?

**Comment by rafaelsaback** [comment]
Added a missing commit here.

@saarnold, I'm positive about this suggestion, do you see any significative drawback here @doudou?

**Comment by doudou** [comment]
Yes, I'm strongly against.

If you generate efforts, don't add a component that does nothing with it -- adding complexity on both sides (the component will become more complex than required)

**Pull request merged**